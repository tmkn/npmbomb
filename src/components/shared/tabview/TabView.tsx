/** @jsx jsx */
import { jsx, css, Global, SerializedStyles } from "@emotion/core";
import React, { useState, useEffect, useContext, useRef } from "react";
import { fromEvent, combineLatest } from "rxjs";
import { merge, filter, debounceTime } from "rxjs/operators";

import { mq, textColor, serifFont, secondaryColorLight, secondaryColor, primaryColorLight, primaryColor, primaryColorDark } from "../../../css";

interface ITabContext {
    activeTab: ITab | null;
    setActiveTab: (tab: ITab) => void;
}

const TabViewContext = React.createContext<ITabContext>({
    activeTab: null,
    setActiveTab: () => {}
});

export interface ITab {
    header: string;
    content: JSX.Element | null;
}

interface ITabView {
    tabs: ITab[];
}

const tabViewStyle = css({
    [mq[0]]: {
        display: `flex`,
        flexDirection: `column`
    }
});

const contentStyle = css({
    [mq[0]]: {
        padding: `2rem 0`
    }
});

export const TabView: React.FC<ITabView> = ({ tabs }) => {
    const [activeTab, setActiveTab] = useState<ITab>(tabs[0]);

    return (
        <TabViewContext.Provider value={{ activeTab, setActiveTab }}>
            <div css={tabViewStyle}>
                <TabHeaders tabs={tabs} />
                <div css={contentStyle}>{activeTab.content}</div>
            </div>
        </TabViewContext.Provider>
    );
};

const tabHeadersStyle = css({
    [mq[0]]: {
        display: `flex`,
        height: `2rem`,
        ">div": {
            display: `flex`,
            justifyContent: `center`,
            fontSize: `1.5rem`,
            fontFamily: serifFont,
            flex: 1,
            alignSelf: `center`,
            cursor: `pointer`,
            "&:hover": {
                backgroundColor: `#e8eaf6`
            }
        }
    }
});

const lineStyle = css({
    [mq[0]]: {
        position: `relative`,
        height: `1px`,
        background: primaryColorDark
    }
});

const TabHeaders: React.FC<ITabView> = ({ tabs }) => {
    const { activeTab, setActiveTab } = useContext(TabViewContext);
    const lineRef = useRef<HTMLDivElement>(null);
    const [highlightWidth, setHighlightWidth] = useState<number>(0);

    const headerTabs: JSX.Element[] = tabs.map((tab, i) => {
        const isActive: boolean = tab.header === activeTab?.header;
        const style = css({
            [mq[0]]: {
                color: isActive ? primaryColor : secondaryColor
            }
        });
        const onTabClick = () => {
            if (lineRef.current) {
                const { width } = lineRef.current.getBoundingClientRect();

                lineRef.current.style.width = `${100 / tabs.length}%`;
                lineRef.current.style.left = `${(100 / tabs.length) * i}%`;
                setHighlightWidth(width / tabs.length);
            }
            setActiveTab(tab);
        };

        return (
            <div key={i} onClick={onTabClick}>
                <span css={style}>{tab.header}</span>
            </div>
        );
    });

    useEffect(() => {
        console.log(lineRef.current);
        if (lineRef.current) {
            const { width } = lineRef.current.getBoundingClientRect();

            lineRef.current.style.width = `${100 / tabs.length}%`;
            setHighlightWidth(width / tabs.length);
        }

        const handleResize = () => {
            if (lineRef.current) {
                const { width } = lineRef.current.getBoundingClientRect();

                lineRef.current.style.width = `${100 / tabs.length}%`;
                setHighlightWidth(width / tabs.length);
            }
        };
        window.addEventListener("resize", handleResize);

        return () => {
            window.removeEventListener("resize", handleResize);
        };
    }, []);

    return (
        <React.Fragment>
            <div css={tabHeadersStyle}>{headerTabs}</div>
            <div ref={lineRef} css={lineStyle} />
        </React.Fragment>
    );
};

const tabs: ITab[] = [
    { header: `Summary`, content: <div>info content</div> },
    { header: `Tree`, content: <div>tree content</div> }
];

export const TestTabView: React.FC = () => <TabView tabs={tabs} />;
