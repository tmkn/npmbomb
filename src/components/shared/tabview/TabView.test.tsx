import React from "react";
import { render, fireEvent, waitForElement } from "@testing-library/react";

import { TabView, ITab } from "./TabView";

describe("<TabView />", () => {
    test("matches snapshot", () => {
        const tabs: ITab[] = [
            { header: `Tab 1`, content: <div>tab 1 content</div> },
            { header: `Tab 2`, content: <div>tab 2 content</div> }
        ];
        const { asFragment } = render(<TabView tabs={tabs} />);

        expect(asFragment()).toMatchSnapshot();
    });

    test("Changes tab on click", async () => {
        const tabHeader1 = `Tab 1`;
        const tabHeader2 = `Tab 2`;
        const tabContent1 = `tab 1 content`;
        const tabContent2 = `tab 2 content`;
        const tabs: ITab[] = [
            { header: tabHeader1, content: <div>{tabContent1}</div> },
            { header: tabHeader2, content: <div>{tabContent2}</div> }
        ];
        const { container, getByText, findByText, queryByText } = render(<TabView tabs={tabs} />);

        await findByText(tabHeader1);
        const tabHeader2El = await findByText(tabHeader2);
        await findByText(tabContent1);
        expect(queryByText(tabContent2)).not.toBeInTheDocument();

        //click on 2nd tab
        fireEvent.click(tabHeader2El);
        await findByText(tabContent2);
        expect(queryByText(tabContent1)).not.toBeInTheDocument();
    });
});
