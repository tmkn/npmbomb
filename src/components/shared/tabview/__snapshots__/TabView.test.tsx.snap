// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TabView /> matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="css-1788elh"
  >
    <div
      class="css-zxjiyd-TabHeaders"
    >
      <div
        tabindex="0"
      >
        <span
          class="css-gvuyi4"
        >
          Tab 1
        </span>
      </div>
      <div
        tabindex="0"
      >
        <span
          class="css-qmn0xh"
        >
          Tab 2
        </span>
      </div>
    </div>
    <div
      class="css-1036lci-TabHeaders"
      style="width: 50%;"
    />
    <div
      class="css-1lr5315"
    >
      <div>
        tab 1 content
      </div>
    </div>
  </div>
</DocumentFragment>
`;
